pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/LeaderService.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.SubscriptionType;
+import org.apache.pulsar.client.impl.ConsumerImpl;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Slf4j
+public class LeaderService implements AutoCloseable, ConsumerEventListener {
+
+    private final String consumerName;
+    private final FunctionRuntimeManager functionRuntimeManager;
+    private final ErrorNotifier errorNotifier;
+    private final SchedulerManager schedulerManager;
+    private ConsumerImpl<byte[]> consumer;
+    private final WorkerConfig workerConfig;
+    private final PulsarClient pulsarClient;
+    private final AtomicBoolean isLeader = new AtomicBoolean(false);
+
+    static final String COORDINATION_TOPIC_SUBSCRIPTION = "participants";
+
+    private static String WORKER_IDENTIFIER = "id";
+    
+    public LeaderService(WorkerService workerService,
+                         PulsarClient pulsarClient,
+                         FunctionRuntimeManager functionRuntimeManager,
+                         SchedulerManager schedulerManager,
+                         ErrorNotifier errorNotifier) {
+        this.workerConfig = workerService.getWorkerConfig();
+        this.pulsarClient = pulsarClient;
+        this.functionRuntimeManager = functionRuntimeManager;
+        this.schedulerManager = schedulerManager;
+        this.errorNotifier = errorNotifier;
+        consumerName = String.format(
+                "%s:%s:%d",
+                workerConfig.getWorkerId(),
+                workerConfig.getWorkerHostname(),
+                workerConfig.getWorkerPort()
+        );
+
+    }
+
+    public void start() throws PulsarClientException {
+        // the leaders service is using a `coordination` topic for leader election.
+        // we don't produce any messages into this topic, we only use the `failover` subscription
+        // to elect an active consumer as the leader worker. The leader worker will be responsible
+        // for scheduling snapshots for FMT and doing task assignment.
+        consumer = (ConsumerImpl<byte[]>) pulsarClient.newConsumer()
+                .topic(workerConfig.getClusterCoordinationTopic())
+                .subscriptionName(COORDINATION_TOPIC_SUBSCRIPTION)
+                .subscriptionType(SubscriptionType.Failover)
+                .consumerEventListener(this)
+                .property(WORKER_IDENTIFIER, consumerName)
+                .consumerName(consumerName)
+                .subscribe();
+
+    }
+
+    @Override
+    public void becameActive(Consumer<?> consumer, int partitionId) {
+        if (isLeader.compareAndSet(false, true)) {
+            log.info("Worker {} became the leader.", consumerName);
+            try {
+                // trigger read to the end of the topic and exit
+                // Since the leader can just update its in memory assignments cache directly
+                functionRuntimeManager.stopReadingAssignments();
```

Comment:
Can you instead call functionRunTimeManager.acquireLeadership() and functionRunTimeManager.giveupLeadership()

Reply:
doesn't make to call "schedulerManager.initialize();" there or add the SchedulerManager as a dependency in FunctionRuntimeManager just for this